### PR TITLE
CUMULUS-1311: Executions queued via queue-granules should not exceeed max

### PR DESCRIPTION
**Pull request by markdboyd**
**Summary:** Summary of changes

Addresses [CUMULUS-1311: As a user, I should be able to limit the number of running executions started from a queue](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1311)

## Changes

* Add test workflow for queuing HelloWorld workflows
* Add integration test that number of running executions queued via queue-granules does not exceed maximum
* Minor refactors to queue-granules and queue-pdrs
* Unit tests for queue-granules and queue-granules that correct message output is written for a queue

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Integration tests

### CODE REVIEW

File: example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
+'use strict';
+
+const fs = require('fs-extra');
+
+const {
+  models: {
+    Collection,
+    Provider
+  }
+} = require('@cumulus/api');
```
Comment (mennovandiermen) [inline example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js]: As long as you're not importing more from `api`, could pull out a level here:
```
const {
  Collection,
  Provider
} = require('@cumulus/api/models');
```

File: example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+    it('has expected count of queued executions', () => {
+      expect(executionArns.length).toEqual(numberOfGranules);
+    });
+
+    it('is only running the maximum number of executions', async () => {
+      // Wait until at least one of the executions has started
+      // TODO: can't rely on first execution ARN being started first
+      await waitForExecutionExists(executionArns[0], 5, 60);
```
Comment (ifestus) [inline example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js]: Probably naive, but would it work to continuously query the `executionExists` like below and just make sure that the # of executions never tops the max?

File: example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+      // Get the state for all of the queued execution ARNs.
+      const getExecutionStates = executionArns
```
Comment (ifestus) [inline example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js]: Probably naive, but would it work to continuously query the executionExists like below and just make sure that the # of executions never tops the max?

Comment (ifestus) [reply example/spec/parallel/queueMaxExecutions/QueueMaxExecutionsSpec.js]:
```
while ("RUNNING" in allExecutionStatus) {
  wait
  allExecutionStatus = ...
}

File: packages/ingest/queue.js [hunk 4]
```diff
@@ -21,37 +22,64 @@ async function getMessageFromTemplate(templateUri) {
...
+ * a PDR parse message
+ * @param {Object} params.provider - the provider config to be attached to the message
+ * @param {Object} params.collection - the collection config to be attached to the
  *   message
- * @param {string} parentExecutionArn - parent workflow execution arn to add to the message
+ * @param {string} params.parentExecutionArn - parent workflow execution arn to add to the message
  * @returns {Promise} - resolves when the message has been enqueued
  */
-async function enqueueParsePdrMessage(
+async function enqueueParsePdrMessage({
```
Comment (ifestus) [inline packages/ingest/queue.js]: good change :+1:

File: example/spec/parallel/startSF/startSfSpec.js [hunk 5]
```diff
@@ -112,4 +166,202 @@ describe('the sf-starter lambda function', () => {
       expect(executions.length).toBe(messagesConsumed);
     });
   });
+
+  xdescribe('when provided a queue with a maximum number of executions', () => {
```
Comment (markdboyd) [inline example/spec/parallel/startSF/startSfSpec.js]: Given the complications with integration testing, I'm going to disable these tests for now but leave them in. I'll pick them back up as part of 812

File: packages/ingest/queue.js [hunk 6]
```diff
@@ -1,111 +1,116 @@
...
+  const message = buildQueueMessageFromTemplate({
+    collection,
+    messageTemplate,
+    parentExecutionArn,
+    payload,
+    provider,
+    queueName
+  });
+
+  if (pdr) message.meta.pdr = pdr;
```
Comment (ifestus) [inline packages/ingest/queue.js]: The fact that `pdr` is a one-off that isn't taken care of by previous logic is somewhat alarming to me - only because I'm worried that there are others. Are there any other values that you can think of that we'd need to add ad-hoc logic for?

Integration tests probably would catch any missing situations...

Comment (ifestus) [reply packages/ingest/queue.js]: nvm this question doesn't make sense.
